### PR TITLE
Backport #27310 Completely revamp the OTTF console code

### PR DESCRIPTION
**Pull request by pamaury**
Backport #27310 Completely revamp the OTTF console code. Depends on #27737

**Note:** this is a straight backport with conflict fixing, notably due to the DT changes on master. However, this PR does not fully convert some aspects to the DT and I will defer that to a further PR. Probably the biggest change would be to replace the `base_addr` attribute in the OTTF test config by a `dt_instance_id_t`.

This should go a long toward unblocking #27619 since now all the UART/SPI_DEVICE specific code is its own file.

### CODE REVIEW

**Comment by pamaury** [comment]
TODO: document more `kSpiDeviceMaxFramePayloadSizeBytes`, see https://github.com/lowRISC/opentitan/pull/27737#discussion_r2260357586

File: sw/device/lib/testing/test_framework/ottf_console.h [hunk 1]
```diff
@@ -43,25 +87,50 @@ void ottf_console_init(void);
...
+ * @param size Length of the staging buffer.
+ */
+void ottf_console_set_buffering(ottf_console_t *console, bool enabled,
+                                char *buffer, size_t size);
 
 /**
  * Manage flow control by inspecting the OTTF console device's receive FIFO.
  *
- * @param uart A UART handle.
+ * @param uart A Pointer to the console.
```
Comment (AlexJones0) [inline sw/device/lib/testing/test_framework/ottf_console.h]: (also present on `earlgrey_1.0.0`)
```suggestion
 * @param console A pointer to the console.
```

File: sw/device/lib/testing/test_framework/ottf_console.h [hunk 2]
```diff
@@ -43,25 +87,50 @@ void ottf_console_init(void);
 /**
  * Configures the given UART to be used by the OTTF console.
  *
+ * @param console Console pointer
```
Comment (AlexJones0) [inline sw/device/lib/testing/test_framework/ottf_console.h:90]: Nit: it would be good to document here (and likewise on `ottf_console_configure_spi_device`) that these functions expect to be called only when `console.type == kOttfConsoleUart` or `kOttfConsoleSpiDevice` respectively, and that otherwise calling them may be unsafe. It might be nice to note this on any of the relevant `ottf_{}_spi_{}` and `ottf_{}_uart_{}` functions with external linkage?

File: sw/device/lib/testing/test_framework/ottf_console.h [hunk 3]
```diff
@@ -107,21 +197,31 @@ uint32_t ottf_console_get_flow_control_irqs(void);
  * @param[out] buf A pointer to the location where the data should be stored.
  * @return The number of bytes actually received from the host.
  */
-size_t ottf_console_spi_device_read(size_t buf_size, uint8_t *const buf);
+size_t ottf_console_spi_device_read(ottf_console_t *console, size_t buf_size,
```
Comment (AlexJones0) [inline sw/device/lib/testing/test_framework/ottf_console.h:201]: Nit: function is missing the `@param console` documentation (also on `earlgrey_1.0.0`)

File: sw/device/lib/testing/test_framework/ottf_console_internal.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ */
+enum {
+  kSpiDeviceReadBufferSizeBytes =
+      SPI_DEVICE_PARAM_SRAM_READ_BUFFER_DEPTH * sizeof(uint32_t),
+  kSpiDeviceFrameHeaderSizeBytes = 12,
+  kSpiDeviceBufferPreservedSizeBytes = kSpiDeviceFrameHeaderSizeBytes,
+  kSpiDeviceMaxFramePayloadSizeBytes = kSpiDeviceReadBufferSizeBytes -
+                                       kSpiDeviceFrameHeaderSizeBytes -
+                                       kSpiDeviceBufferPreservedSizeBytes - 4,
+  kSpiDeviceFrameMagicNumber = 0xa5a5beef,
```
Comment (AlexJones0) [inline sw/device/lib/testing/test_framework/ottf_console_internal.h:29]: Nit: moving this into the enum technically changes the type of `kSpiDeviceFrameMagicNumber` to an `int` from a `uint32_t`. This doesn't seem to be a problem, since we always mask the shifted bits in the code that uses it (so the sign bits inserted as a result of the arithmetic right shift never materialise as an issue), but this might not be obvious/expected behaviour if other code is introduced that uses this value.

File: sw/device/lib/testing/test_framework/ottf_console_spi.c [hunk 5]
```diff
@@ -87,7 +86,7 @@ static size_t spi_device_send_frame(void *data, const char *buf, size_t len) {
   }
   // Add the frame number.
   for (size_t i = 0; i < 4; ++i) {
-    frame_header_bytes[i + 4] = (spi_device_frame_num >> (i * 8)) & 0xff;
```
Comment (AlexJones0) [inline sw/device/lib/testing/test_framework/ottf_console_spi.c]: Aside (comment on the original code, unrelated to this PR's changes): it might be preferable for the code here to keep one mutable index into `frame_header_bytes` to ensure that any changes to the frame format do not result in accidental overwriting, e.g.:

```c
  size_t header_index = 0;
  // Add the magic bytes.
  for (size_t i = 0; i < 4; ++i) {
    frame_header_bytes[header_index++] = (kSpiDeviceFrameMagicNumber >> (i * 8)) & 0xff;
  }
  // Add the frame number.
  for (size_t i = 0; i < 4; ++i) {
    frame_header_bytes[header_index++] = (console->data.spi.frame_num >> (i * 8)) & 0xff;
  }
  // etc.
```

File: sw/device/lib/testing/test_framework/ottf_console.c [hunk 6]
```diff
@@ -75,57 +76,92 @@ void ottf_console_init(void) {
...
-    }
-    spi_buf_end = 0;
+  if (console->buffered && console->buf_end > 0) {
+    written_len = console->sink(console, console->buf, console->buf_end);
+    size_t lost = console->buf_end - written_len;
+    console->buf_end = 0;
+    return lost;
   }
-  return OK_STATUS((int32_t)written_len);
+  return 0;
```
Comment (AlexJones0) [inline sw/device/lib/testing/test_framework/ottf_console.c]: Nit: the code here is kind of awkward. I would either return `written_len` at the end here, or just move the declaration of `written_len` into the if branch.

File: sw/device/lib/testing/test_framework/ottf_console.h [hunk 7]
```diff
@@ -44,6 +44,14 @@ typedef struct ottf_console {
   sink_func_ptr sink;
   /* Function pointer to a function that retrieves a single character. */
   status_t (*getc)(void *);
+  /* Enable SW buffering. */
+  bool buffered;
+  /** Staging buffer. */
+  char *buf;
+  /** Staging buffer size. */
+  size_t buf_sz;
```
Comment (AlexJones0) [inline sw/device/lib/testing/test_framework/ottf_console.h]: Nit (no need to change for this PR) - `sz` is a bit of a needless abbreviation, better to be clear with `buf_size`.

File: sw/device/lib/testing/test_framework/ottf_console.c [hunk 8]
```diff
@@ -75,57 +76,92 @@ void ottf_console_init(void) {
...
+                                          const char *buf, size_t len) {
+  if (len > console->buf_sz) {
     // Flush and skip the staging buffer if the payload is already the max size.
-    TRY(ottf_console_flushbuf(io));
-    size_t written_len = console->sink(io, buf, len);
-    if (len != written_len) {
-      return DATA_LOSS((int32_t)(len - written_len));
+    size_t lost = ottf_console_flush(console);
+    if (lost > 0) {
+      return 0;
```
Comment (AlexJones0) [inline sw/device/lib/testing/test_framework/ottf_console.c]: Why the change from returning a `status_t` to a `size_t`? It looks like previously the function would return either `Ok(length written)` or `Err(Data loss = bytes lost)`, but now it just returns either the `length written`, or `0` to indicate a failure (which is now indistinguishable from an empty write)? I see that you've wrapped this logic in `ottf_console_flushbuf` but it seems that now buffered writes which partially succeed lose the context of how many chars were successfully written.

File: sw/device/lib/testing/test_framework/ottf_console.h [hunk 9]
```diff
@@ -43,25 +87,50 @@ void ottf_console_init(void);
...
+
+/**
+ * Configures software buffering of the console.
+ *
+ * If the console previously had buffering enabled, disabling buffering will
+ * simply drop the content of the staging buffer. Therefore the content should
+ * be flushed before disabling buffering.
+ *
+ * @param console Pointer to the console to configure
+ * @param enable Enable/Disable buffering.
```
Comment (AlexJones0) [inline sw/device/lib/testing/test_framework/ottf_console.h]: (also on `earlgrey_1.0.0`)
```suggestion
 * @param enabled Enable/disable buffering.
```

File: sw/device/lib/testing/test_framework/ottf_console.h [hunk 10]
```diff
@@ -74,11 +143,19 @@ status_t ottf_console_flow_control(const dif_uart_t *uart,
  *
  * This function configures UART interrupts at the PLIC and enables interrupts
  * at the CPU.
+ *
+ * WARNING Control requires IRQ dispatching. The main console will automatically
+ * dispatch control IRQs on calls to `ottf_console_flow_control_isr`. If you
+ * want to dispatch control flow IRQs for non-main console, you must call
+ * manually call `ottf_console_uart_flow_control_isr` on the console from your
```
Comment (AlexJones0) [inline sw/device/lib/testing/test_framework/ottf_console.h]: (also on `earlgrey_1.0.0`)
```suggestion
 * want to dispatch control flow IRQs for non-main consoles, you must
 * manually call `ottf_console_uart_flow_control_isr` on the console from your
```

File: sw/device/lib/testing/test_framework/ottf_console_uart.c [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+  if (rx) {
+    manage_flow_control(console, kOttfConsoleFlowControlAuto);
+    CHECK_DIF_OK(dif_uart_irq_acknowledge(uart, kDifUartIrqRxWatermark));
+    return true;
+  }
+  return false;
+}
+
+// The public API has to save and restore interrupts to avoid an
+// unexpected write to the global `console->data.uart.flow_control_state`.
```
Comment (AlexJones0) [inline sw/device/lib/testing/test_framework/ottf_console_uart.c]: (also on `earlgrey_1.0.0`)
```suggestion
// unexpected write to `console->data.uart.flow_control_state`.
```

**Review by AlexJones0 — COMMENTED** [review]
Thanks for backporting this @pamaury, it is definitely an improvement. I've left a few nits (feel free to leave these for follow-up PRs if you want to ease backporting efforts) and a couple of questions.

File: sw/device/lib/testing/test_framework/ottf_console.h [hunk 12]
```diff
@@ -31,10 +33,31 @@ typedef enum ottf_console_flow_control {
...
+typedef struct ottf_console {
+  /** Console type. */
+  ottf_console_type_t type;
+  /** Auxiliary data, per console type */
+  union {
+    /** UART data. */
+    struct {
+      /** DIF handle. */
+      dif_uart_t dif;
+    } uart;
```
Comment (engdoreis) [inline sw/device/lib/testing/test_framework/ottf_console.h:64]: Why wrap the diff handle in a struct? Unless the structs will have other fields I would suggest removing the extra layer.

File: sw/device/lib/testing/test_framework/ottf_console.c [hunk 13]
```diff
@@ -75,57 +76,92 @@ void ottf_console_init(void) {
...
 
-status_t ottf_console_flushbuf(void *io) {
-  ottf_console_t *console = io;
+void ottf_console_set_buffering(ottf_console_t *console, bool enabled,
+                                char *buffer, size_t size) {
+  console->buffered = enabled;
+  if (enabled) {
+    console->buf = buffer;
+    console->buf_sz = size;
+    console->buf_end = 0;
```
Comment (engdoreis) [inline sw/device/lib/testing/test_framework/ottf_console.c]: Nit. This line repeats in both arms and can the extracted.

File: sw/device/lib/testing/test_framework/ottf_console.h (hunk 2 above)
Comment (pamaury) [reply sw/device/lib/testing/test_framework/ottf_console.h:90]: That's a good point, actually I will move the setting of `console.type` to those functions since it makes more sense. I will also add a bunch of `CHECK` on the types in public API functions.

File: sw/device/lib/testing/test_framework/ottf_console_internal.h (hunk 4 above)
Comment (pamaury) [reply sw/device/lib/testing/test_framework/ottf_console_internal.h:29]: That's true (although it was done in a previous PR). There was a good reason to move the other to an enum (use them in a type definition) but I could move this back to a constant variable.

File: sw/device/lib/testing/test_framework/ottf_console_spi.c (hunk 5 above)
Comment (pamaury) [reply sw/device/lib/testing/test_framework/ottf_console_spi.c]: Added to #27852

File: sw/device/lib/testing/test_framework/ottf_console.c (hunk 8 above)
Comment (pamaury) [reply sw/device/lib/testing/test_framework/ottf_console.c]: Fair point, I change it back. The reason I changed it to a `size_t` because we do not have proper error tracking in the console code anyway. For example the sinking functions return a `size_t`, not a `status_t`.

Another point to note that is retuning data with an error code (e.g. `DATA_LOSS`) is essentially best effort. In practice it's stored on only 11-bit in place of the line number. For a buffer size this might even be too small.

File: sw/device/lib/testing/test_framework/ottf_console.h (hunk 12 above)
Comment (pamaury) [reply sw/device/lib/testing/test_framework/ottf_console.h:64]: In #27747, this becomes a per-device-type (e.g. UART, SPI) structure defined in another header.

**Comment by pamaury** [comment]
I have restructured some of the code to return `status_t` everwhere. Hopefully I didn't break anything in the process
